### PR TITLE
Special fixes

### PR DESCRIPTION
**Pull request by steppi**
This PR makes a few small fixes to the special functions that currently exist in mparray. 

The fixes here are:

1. Change the signature of `mparray.gamminc` to match that of `scipy.special.gammainc`. That of `mparray.gammainc` is currently `gammainc(x, a)` and that of `scipy.special.gammainc` is `gamma(a, x)`, where the meaning of `x` and `a` is swapped. This was an easy mixup to make because `mparray` currently follows `mpmath` which also gives the opposite meanings to these variable names compared to SciPy. I think the variable names in `mpmath` are just an implementation detail.
2. Handle boundary and out of bounds cases in `ndtri`. It currently errors out for `x = 0` and for `x < 0 or x > 1`. It should return `-inf` and `nan` respectively in these cases.
3. Handle out of bounds cases in `betainc`.
4. Fix bug in `cosm1` causing negative `extra_dps` for large `x`.
5. Make `cosm1` handle catastrophic cancellation near nonzero integer multiples of $2\pi$.
6. Fix typo in `boxcox`, `mp.lmbda(x)` which should have been `mp.log(x)`.
7. Handle out of bounds cases for `boxcox` and `boxcox1p`.
8. Fix precision issues for `boxcox1p` near `x = 0`, by setting higher working dps when needed.

I've also changed increases of working dps to use the `mp.workdps` context manager.

There will be followups to add more functions which already exist in mpmath, and to implement more functions which do not currently exist in mpmath.

### CODE REVIEW

File: mparray/special/_special.py [hunk 1]
```diff
@@ -31,22 +31,27 @@
...
-    return mp.gammainc(x, a=0, b=a, regularized=True)
+def gammainc(a, x):
+    return mp.gammainc(a, a=0, b=x, regularized=True)
 
 
 @vectorize
-def gammaincc(x, a):
-    return mp.gammainc(x, a=a, b=mp.inf, regularized=True)
+def gammaincc(a, x):
+    return mp.gammainc(a, a=x, b=mp.inf, regularized=True)
```
Comment (mdhaber) [inline mparray/special/_special.py:40]: Oops. I was mass-producing these so I suppose I equated `z` with `x`, etc. 

Just to be clear: if the function didn't *work* correctly - if I had the arguments swapped instead of just swapping the names - the tests would have caught it, right?

File: mparray/special/_special.py [hunk 2]
```diff
@@ -31,22 +31,27 @@
...
 
 
 @vectorize
 def ndtri(x):
+    if x == 0:
+        return -mp.inf
+    if x == 1:
+        return mp.inf
+    if x < 0 or x > 1:
+        return mp.nan
```
Comment (mdhaber) [inline mparray/special/_special.py:50]: Good call. An improvement to the infrastructure might be to have a standard way to define the domain, what to return for edge cases, and what to return for out-of-bounds cases. Then we could test those cases.

File: mparray/special/_special.py [hunk 3]
```diff
@@ -31,22 +31,27 @@
...
+
     extra_dps = int(mp.ceil(-mp.log10(x)))
-    mp.dps += extra_dps
-    res = mp.sqrt(2) * mp.erfinv(2 * x - mp.one)
-    mp.dps -= extra_dps
-    return res
+    with mp.workdps(mp.dps + extra_dps):
```
Comment (mdhaber) [inline mparray/special/_special.py:53]: Cool. I thought I looked for something like this but I guess I didn't look carefully enough.

File: mparray/special/_special.py [hunk 4]
```diff
@@ -65,6 +70,10 @@ def betaln(x, y):
 
 @vectorize
 def betainc(a, b, x):
+    if x < 0 or x > 1:
+        # The mpmath betainc implementation is defined on entire real line,
+        # (and the complex plane). We want to match scipy.special.betainc.
+        return mp.nan
```
Comment (mdhaber) [inline mparray/special/_special.py:76]: Hmm. I suppose when writing the standard we should have a minimum standard for the domain and what to do out-of-domain, but is there harm for extending the domain? Perhaps there are different extensions, and we'd have to define which to follow?

File: mparray/special/_special.py [hunk 5]
```diff
@@ -90,12 +99,15 @@ def xlog1py(x, y):  # needs accuracy review
 
 @vectorize
 def cosm1(x):
+    if x == 0:
+        return mp.zero
```
Comment (mdhaber) [inline mparray/special/_special.py:103]: Just a shortcut? It seems to give the right result without this.

File: mparray/special/_special.py [hunk 6]
```diff
@@ -109,20 +121,30 @@ def expit(x):  # needs accuracy review
...
+
+
+@vectorize
+def boxcox(x, lmbda):
+    return _boxcox_scalar(x, lmbda)
 
 
-def boxcox1p(x, lmbda):  # needs accuracy review
-    return boxcox(mp.one + x, lmbda)
+def boxcox1p(x, lmbda):
```
Comment (mdhaber) [inline mparray/special/_special.py:142]: Doesn't this need to be vectorized, then?

File: mparray/special/_special.py [hunk 7]
```diff
@@ -109,20 +121,30 @@ def expit(x):  # needs accuracy review
...
+    if x < - 1:
+        return mp.nan
```
Comment (mdhaber) [inline mparray/special/_special.py:144]: Isn't this covered by `_boxcox_scalar`?

File: mparray/special/_special.py [hunk 8]
```diff
@@ -109,20 +121,30 @@ def expit(x):  # needs accuracy review
...
     """
     y = (x**lmbda - 1) / lmbda  if lmbda != 0
         log(x)                  if lmbda == 0
     """
+    if x < 0:
+        return mp.nan
     if lmbda != 0:
         return mp.powm1(x, lmbda) / lmbda
     else:
-        return mp.lmbda(x)
```
Comment (mdhaber) [inline mparray/special/_special.py:121]: Good catch!

**Review by mdhaber — APPROVED** [review]
Basically all great; just a few comments. Feel free to resolve those without a question mark.

File: mparray/special/_special.py (hunk 1 above)
Comment (steppi) [reply mparray/special/_special.py:40]: Yeah, I think the tests would have caught it.

File: mparray/special/_special.py (hunk 2 above)
Comment (steppi) [reply mparray/special/_special.py:50]: Yeah. That’s a good idea I think.

File: mparray/special/_special.py (hunk 4 above)
Comment (steppi) [reply mparray/special/_special.py:76]: I can’t remember exactly why I restricted the domain and am not at my computer now to check. I know I did it after implementing `betaincinv`, which isn’t included in this PR. We can drop the domain restriction here I think.

File: mparray/special/_special.py (hunk 5 above)
Comment (steppi) [reply mparray/special/_special.py:103]: Yeah. I think this was just done mindlessly out of habit

File: mparray/special/_special.py (hunk 6 above)
Comment (steppi) [reply mparray/special/_special.py:142]: Good catch on this and below. What happened is that I copied this over from a branch where I implemented the inverses of the boxcox functions. There I also have a `_boxcox1p_scalar` which is used internally by `inv_boxcox1p`. This is just the scalar version which is why it has the unnecessary check and lacks vectorization.

File: mparray/special/_special.py (hunk 5 above)
Comment (mdhaber) [reply mparray/special/_special.py:103]:
```suggestion
```

File: mparray/special/_special.py (hunk 6 above)
Comment (mdhaber) [reply mparray/special/_special.py:142]: OK. Will you add it, or do you want it merged as-is?

File: mparray/special/_special.py (hunk 4 above)
Comment (mdhaber) [inline mparray/special/_special.py:76]:
```suggestion
```

File: mparray/special/_special.py (hunk 6 above)
Comment (steppi) [reply mparray/special/_special.py:142]: I’ll add the inverses in a future PR. This boxcox1p needs to be fixed before merging though.

**Comment by steppi** [comment]
Oops. I think might have jumped the gun getting rid of directly handling `x = 0` for `cosm1`. When I get home I’ll check why that was needed.

**Comment by mdhaber** [comment]
Locally, 
```python3
mp.cos(0) - mp.one  # mpf('0.0')
mp.dps = 100
mp.cos(0) - mp.one  # mpf('0.0')
mp.cos(mp.zero) - mp.one  # mpf('0.0')
```

**Comment by steppi** [comment]
> Locally,
> 
> ```python
> mp.cos(0) - mp.one  # mpf('0.0')
> mp.cos(mp.zero) - mp.one  # mpf('0.0')
> mp.dps = 100
> mp.cos(0) - mp.one  # mpf('0.0')
> mp.cos(mp.zero) - mp.one  # mpf('0.0')
> ```



> Locally,
> 
> ```python
> mp.cos(0) - mp.one  # mpf('0.0')
> mp.cos(mp.zero) - mp.one  # mpf('0.0')
> mp.dps = 100
> mp.cos(0) - mp.one  # mpf('0.0')
> mp.cos(mp.zero) - mp.one  # mpf('0.0')
> ```

It’s the extra_dps calculation. You get an error because of the log(0).

**Comment by mdhaber** [comment]
I see.
